packages/vite-plugin-cloudflare/src/index.ts

```diff
@@ -407,67 +410,53 @@ export function cloudflare(pluginConfig: PluginConfig = {}): vite.Plugin[] {
 				// Skip this whole plugin if we are in preview mode
 				return !env.isPreview;
 			},
-			config() {
-				// Configure Vite with the Node.js polyfill aliases
-				// We have to do this across the whole Vite config because it is not possible to do it per Environment.
-				// These aliases are to virtual modules that then get Environment specific handling in the resolveId hook.
+			applyToEnvironment(environment) {
+				const workerConfig = getWorkerConfig(environment.name);
+				return isNodeCompat(workerConfig);
+			},
+			// We need the resolver from this plugin to run before built-in ones, otherwise Vite's built-in
+			// resolver will try to externalize the Node.js module imports (e.g. `perf_hooks` and `node:tty`)
+			// rather than allowing the resolve hook here to alias then to polyfills.
+			enforce: "pre",
+			configEnvironment() {
 				return {
 					resolve: {
-						alias: getNodeCompatAliases(),
+						builtins: getNodeCompatExternals(),
+					},
+					optimizeDeps: {
+						// We don't want to follow any `node:*` type imports when optimizing deps,
+						// whether or not they are built-in to workerd or polyfilled here.
+						// The ones that are not built-in will just get resolved to their polyfill on demand.
+						exclude: [...nodeBuiltInModules],
+						// But we do want to ensure that all the possible entry-points for the polyfills get pre-bundled
+						// so that they are ready for the first request to the dev server.
+						include: [...getNodeCompatEntries()],
```

Comment:
Does `include` override `exclude` here? All Node builtins are excluded, including those that are unprefixed and those that are prefixed with `node:` and then `include` adds the polyfilled builtins back in? If so, I think the comments could make this clearer.

Reply:
Include is the list of packages that should be pre-bundled.
Exclude is the list of imports that should not be followed when doing the import analysis.
I am not quite sure what happens when there is overlap.
But in our case here, there is no overlap. The entries are all of the form `unenv/runtime/...` or `@cloudflare/unenv-preset/runtime/...`.
I'll try to make the comments clearer.

Reply:
Sorry, I should have deleted this comment after looking at the rest of the code because it made sense then.

Reply:
4de9f48d9c6af8eb2319b4cb8ce5f4f892b99c53